src/gui/qgsadvanceddigitizingcanvasitem.cpp

```diff
@@ -37,6 +38,31 @@ void QgsAdvancedDigitizingCanvasItem::paint( QPainter *painter )
   if ( !mAdvancedDigitizingDockWidget->cadEnabled() )
     return;
 
+  painter->setRenderHint( QPainter::Antialiasing );
+  painter->setCompositionMode( QPainter::CompositionMode_Difference );
+
+  // Draw construction guides
+  if ( mAdvancedDigitizingDockWidget->showConstructionGuides() )
+  {
+    if ( QgsVectorLayer *constructionGuidesLayer = mAdvancedDigitizingDockWidget->constructionGuidesLayer() )
+    {
+      QgsFeatureIterator it = constructionGuidesLayer->getFeatures( mMapCanvas->mapSettings().visibleExtent() );
+      QgsFeature feature;
+      painter->setPen( mConstructionGuidesPen );
+      while ( it.nextFeature( feature ) )
+      {
+        QgsPolylineXY line = feature.geometry().asPolyline();
+        QPolygonF polygon( line.size() );
+        for ( int i = 0; i < line.size(); i++ )
+        {
+          polygon[i] = toCanvasCoordinates( line[i] );
+        }
+
+        painter->drawPolyline( polygon );
```

Comment:
I think we could optimise this:

```suggestion
        QgsGeometry geom = feature.geometry();
        geom.mapToPixel( mMapCanvas->getCoordinateTransform() );
        const QPolygonF polygon = geom.asQPolygonF();
        painter->drawPolyline( polygon );
```

Reply:
@nyalldawson , great idea.